### ISSUE TITLE
Feature: a way replace inner function in `spy` mocking 

### ISSUE THREAD

**Issue by CyanChanges** [denoland/std#6234]
**Is your feature request related to a problem? Please describe.**

In an event emitter, I want to replace the handler to throw an error, to verify if the emitter can handle error

**Describe the solution you'd like**

Like `node:test`, you can replace the mock implementation with
```ts
import { mock } from 'node:test'
mocked_fn = mock.fn()
mocked_fn.mock.mockImplementation(()=>{ /* new logic */ })
```

**Describe alternatives you've considered**

make the `fn` in `spy(fn)` a wrapper, which call an external function variable.

**Comment by IgorM867** [comment]
How exactly would this look like in terms of `std`? What do you think of something like that?
```ts
Deno.test("test", () => {
  let cnt = 0;

  function addOne() {
    cnt++;
    return cnt;
  }

  function addTwo() {
    cnt += 2;
    return cnt;
  }

  function throws() {
    throw new Error("error");
  }

  const fn = spy(addOne);

  assertEquals(fn(), 1);
  fn.mockImplementation(addTwo);

  assertEquals(fn(), 3);
  assertEquals(fn(), 5);

  fn.mockImplementation(throws as () => number);
  assertThrows(fn, "error");
});
```

**Comment by CyanChanges** [comment]
> How exactly would this look like in terms of `std`? What do you think of something like that?
> 
> ```ts
> Deno.test("test", () => {
>   let cnt = 0;
> 
>   function addOne() {
>     cnt++;
>     return cnt;
>   }
> 
>   function addTwo() {
>     cnt += 2;
>     return cnt;
>   }
> 
>   function throws() {
>     throw new Error("error");
>   }
> 
>   const fn = spy(addOne);
> 
>   assertEquals(fn(), 1);
>   fn.mockImplementation(addTwo);
> 
>   assertEquals(fn(), 3);
>   assertEquals(fn(), 5);
> 
>   fn.mockImplementation(throws as () => number);
>   assertThrows(fn, "error");
> });
> ```

LGTM
I think could also support a setter, like:
```
fn.implementation = ()={ ... }
```

**Comment by kt3k** [comment]
`spy` function already exposes (readonly) `.original` property. Maybe making it writable property is the easiest way to implement this.

cc @KyleJune What do you think of this kind of addition?

**Comment by CyanChanges** [comment]
> `spy` function already exposes (readonly) `.original` property. Maybe making it writable property is the easiest way to implement this.
> 
> cc @KyleJune What do you think of this kind of addition?

Yes, I checked the code. Just add a setter.
https://github.com/denoland/std/blob/2ad1564436ea96dce8174694c93b9b6541a5f77b/testing/mock.ts#L653-L656

**Comment by KyleJune** [comment]
> > `spy` function already exposes (readonly) `.original` property. Maybe making it writable property is the easiest way to implement this.
> > cc @KyleJune What do you think of this kind of addition?
> 
> Yes, I checked the code. Just add a setter here.
> 
> https://github.com/denoland/std/blob/2ad1564436ea96dce8174694c93b9b6541a5f77b/testing/mock.ts#L653-L656

That original property you linked is on the methodSpy but there is a similar one on functionSpy. The purpose of it is to be able to access the original value of the property. Making it writable and giving the user access to override it would basically turn spys into stubs. It would also make the property confusing since the name implies it would give you the original function being spied on while it actually would be giving you the stub function.

I don't think original should be made writable. If you want to make it possible to turn a spy into a stub or to replace the replacement function of a stub, I would start with changing it so that the fake property on stubs is writable and updating the stub code so that it will call that property instead of the fake const. That would make it possible to swap the fake version of the function with a new fake that you would like the stub to use going forward.

https://github.com/denoland/std/blob/2ad1564436ea96dce8174694c93b9b6541a5f77b/testing/mock.ts#L1106-L1109

Then for making it possible to convert a spy, I'd suggest adding a fake property to spies that defaults to undefined. Then if the user sets it, make the spy call through to the fake instead of the original if the fake property is set. That would also make it possible to detect if a spy has been converted into a stub.

Oh also instead of just making it a writable, I'd actually change it so that the fake variable is declared with let instead of const. Then the fake property descriptor would have a getter that returns the vale and a setter that overwrites it. That way we could have the setter throw an error if the user tries to set it to anything other than undefined or a function.

I don't like the idea of there being 2 different ways to change the implementation used behind the mock. I'd suggest choosing to either make fake property a getter/setter, or add a function like replaceStub. Personally, I'd prefer the fake property getter/setter approach.

**Comment by kt3k** [comment]
@KyleJune I agree with you. Let's not make `original` writable, but add `.fake` to spy and make it writable

**Comment by timreichen** [comment]
How about `replacement` or `substitute` instead of `fake` as a name?

**Comment by gmzacharydovel** [comment]
I've used the sinon patterns back in the day and never really understood the arbitrary complexity of separate fake/stub/spy operations.  If that property gets added, it will be bit easier to mimic the mocking patterns a lot of popular libraries.

**Comment by gmzacharydovel** [comment]
These are my two cents, if there was a `@std/testing/jest` or something similar, it would probably be a lot easier for many projects to migrate to deno.

The below is some code to mimic how jest mocks/spys work that could be copied to a testing.ts for anyone reading this thread..  I'm probably going to create package on jsr at some point

It supports the automatic detruction with the `using` keyword.

```USAGE
const api = { send: () => };

using responseMock = createMock(() => 'result');
console.log(responseMOck.mock.calls)

using apiSpy = createSpy(api, 'send');
```

```CODE
// deno-lint-ignore no-explicit-any
type AnyFunction = (...args: any) => any;
// deno-lint-ignore ban-types
type ParentObject = Object;

type MockCall<T extends AnyFunction> = {
  args: Parameters<T>;
  returns: ReturnType<T>;
};
type MockImplementation<T extends AnyFunction> = (
  ...args: Parameters<T>
) => ReturnType<T>;
type Callable<T extends ParentObject, Key extends keyof T> = T[Key] extends
  AnyFunction ? T[Key]
  : never;
type Spy<T extends ParentObject, Key extends keyof T> = Mock<Callable<T, Key>>;
type Mock<T extends AnyFunction> = T & {
  [Symbol.dispose]: () => void;
  mock: {
    calls: MockCall<T>[];
    callsArgs: MockCall<T>["args"][];
    callsReturns: MockCall<T>["returns"][];
  };
  mockClear: () => void;
  mockImplementation: (implementation: MockImplementation<T>) => void;
  mockRestore: () => void;
};

export const createSpy = <T extends ParentObject, Key extends keyof T>(
  data: T,
  key: Key,
): Spy<T, Key> => {
  const originalFunction = data[key];
  const out = createMock(originalFunction as AnyFunction);
  const originalMockRestore = out.mockRestore;

  data[key] = out as T[Key];

  // Patch mock restore to also restore tye spied attribute
  out.mockRestore = () => {
    data[key] = originalFunction;
    originalMockRestore();
  };

  return out as unknown as Spy<T, Key>;
};

const fallbackImplementation = () => undefined;

export const createMock = <T extends AnyFunction>(
  initialImplementation: MockImplementation<T> =
    (fallbackImplementation as MockImplementation<T>),
): Mock<T> => {
  let implementation = initialImplementation;

  const mock: {
    calls: MockCall<T>[];
    callsArgs: MockCall<T>["args"][];
    callsReturns: MockCall<T>["returns"][];
  } = {
    calls: [],
    callsArgs: [],
    callsReturns: [],
  };

  out.mock = mock;

  // This MUST be a function to inherit "this"
  function out(this: T, ...args: Parameters<T>): ReturnType<T> {
    const returns = implementation.apply(this, args);

    mock.calls.push({ args, returns });
    mock.callsArgs.push(args);
    mock.callsReturns.push(returns);

    return returns;
  }

  (out as Mock<T>)[Symbol.dispose] = (): void => {
    out.mockRestore();
  };

  out.mockImplementation = (newImplementation: T) => {
    implementation = newImplementation;
  };

  out.mockRestore = () => {
    implementation = initialImplementation;
  };

  out.mockClear = () => {
    mock.calls.length = 0;
    mock.callsArgs.length = 0;
    mock.callsReturns.length = 0;
  };

  return out as unknown as Mock<T>;
};
```

This works pretty well on my current project so far.

**Comment by kt3k** [comment]
@gmzacharydovel jest style mock function is available from `@std/expect`

```js
import { expect, fn } from "jsr:@std/expect";
const mock = fn();
mock("foo", 42);
expect(mock).toHaveBeenCalledWith("foo", 42);
```

**Comment by gmzacharydovel** [comment]
Nice! Sorry, I missed it.  I'm still getting used to jsr.  

Might it be worth adding something about fn on the overview?

https://jsr.io/@std/expect/doc/fn/~/fn

**Comment by gmzacharydovel** [comment]
@kt3k  This is quite nice.  

QQ: Is there a plan to add `spyOn` to it? Or is the approach expected to be done differently?

**Comment by kt3k** [comment]
> QQ: Is there a plan to add spyOn to it? Or is the approach expected to be done differently?

`std/expect` generally tries to be compatible with jest's assertions and mock utils. So adding `spyOn` to it makes sense. I created an issue for it https://github.com/denoland/std/issues/6449

Deno core team currently can't work on it, but contributions are very welcome

**Comment by gmzacharydovel** [comment]
@kt3k  I have some extra cycles so I'll take a look at it.
